### PR TITLE
Ensure void elements have closing slash in mermaid svg

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- starts #15660

## Code changes

- [x] transform `<br>` tags into closed `<br/>` when generating the `img` tag containing mermaid SVG
  - > the regular expression will actually find [all void elements](https://developer.mozilla.org/en-US/docs/Glossary/Void_element), and preserves attributes
- [x] test
  - [x] add a test/notebook with examples of all (but one) of the upstream documentation examples
  - [x] add screenshots

## User-facing changes

- users will be able to use expressive more typographic features in mermaid diagrams

## Security/Compatibility considerations
- some implementations (e.g. GitHub) will _not even parse_ diagrams with any void elements (or probably _any_ fancy HTML) _except_ `hr` and `br`
  - currently, in this PR, as in the [mermaid playground](http://mermaid.live), one can do really interesting/outrageous things, for example:
    - the `style="border-color: red"` attribute can be used on an `hr` to change its color
    - `<input type="checkbox" />` and `radio` can be used to do weird things
  - in the playground, these are actually clickable, etc. but sadly not feasible with the (otherwise robust and portable) `` approach
    - though text labels _may_ contain further `img` tags
- no implementation render `<script>` tags
- `onclick="javascript:alert('pwnd')"` attributes _do_ appear to pass through on the playground, and in `4.1.0b1` **if** a user goes through _right-click, open in new tab_ and then interacts with the diagram

## Backwards-incompatible changes

- if we decide to do more aggressive sanitizing, this could break some diagrams/examples in the wild

### CODE REVIEW

**Comment by krassowski** [comment]
Upstream issue: https://github.com/mermaid-js/mermaid/issues/1766

What effect the suggested `flowchart.htmlLabels=false` have on compatibility?

**Comment by bollwyvl** [comment]
Yeah, there's that one (not sure how breaking). I need to take a look at
various other control mechanisms. There's also a sandbox mode, which is
probably worth investigating, though it does target an iframe with html
srcdoc instead of SVG, so does more weird stuff to entities in
foreignobjects. Ending up with an img tag is still the most portable output.

**Comment by bollwyvl** [comment]
> htmlLabels=false

This does not solve the problem for `\n` in node labels.

> `sandbox`

This resulted in strictly _worse_ results e.g. `&gt;br&lt;` as these are "magic" entities, not defined in the SVG spec.

The `<iframe srcdoc>` approach _is_ very attractive, as, for example, HTML (and therefore markdown) images can do icky things like:

```md
This is totally valid markdown:

<img src="/api/shutdown" alt="shutdown"/>
``` 

And pretty much all frontends will render it, including (at least in preview) GitHub

> This is totally valid markdown:
> 
> shutdown

But `marked` communicates over strings, and filters out `iframe`s... we probably _don't_ want to change that posture, today, as they can very much be used for deceptive things, not _just_ their "white hat" security techniques).

**Comment by krassowski** [comment]
> But `marked` communicates over strings, and filters out `iframe`s... we probably don't want to change that posture, today, as they can very much be used for deceptive things, not just their "white hat" security techniques).

Could we use the same splicing approach as we do with MathJax? See:

https://github.com/jupyterlab/jupyterlab/blob/e38e0b8d59b8cba1d559040236a142931c17c42b/packages/rendermime/src/renderers.ts#L323-L330

**Comment by krassowski** [comment]
Kicking this PR to check if the labeller for mermaid now works (https://github.com/jupyterlab/jupyterlab/pull/15717), also FYI once you merge/rebase the tests for mermaid should now run on CI.

**Comment by bollwyvl** [comment]
Thanks. Alas, it might be a tick before I can get back to this...

File: galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png
Comment (krassowski) [inline galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png:1]: This looks amiss? In general I am not a fan of that many snapshots but I am ok with these being here as long as they are isolated to the diagrams alone (no other UI).

File: packages/mermaid/src/manager.ts
Comment (krassowski) [inline packages/mermaid/src/manager.ts:86]: I would love to see this unit tested. For example input `'<br/>'` (which I understand we do not expect from the current version of marked but I guess user or future versions could introduce it) results in `'<br / >'` which strictly speaking is invalid XHTML. Basically:

```js
new DOMParser().parseFromString('<br>'.replace(RE_VOID_ELEMENT, replaceVoidElement), 'image/svg+xml')
```

works, but:

```js
new DOMParser().parseFromString('<br/>'.replace(RE_VOID_ELEMENT, replaceVoidElement), 'image/svg+xml')
```

fails.

**Review by krassowski — CHANGES_REQUESTED** [review]

**Comment by bollwyvl** [comment]
Ok, all of the screenshots are now working (and less sensitive to scroll skew). Will merge upstream, and see if anything pops out, but no further work planned.

File: galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png
Comment (bollwyvl) [reply galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png:1]: Of _all_ the things a screenshot is good at, _diagram renders right_ seems like a legit use case vs all the "whole screen" shots that are sensitive to individual pixel skew. 

At any rate, these are now as isolated as they can be... the "screenshot element" function doesn't handle bigger-than-viewport stuff very well, and there are two examples diagrams that are larger than the viewport.

File: packages/mermaid/src/manager.ts
Comment (bollwyvl) [reply packages/mermaid/src/manager.ts:86]: Sure, will add some coverage. 

I'm semi-disappointed we _still_ don't have any [property-based testing](https://github.com/dubzzz/fast-check) or overarching [frontend coverage](https://github.com/istanbuljs/istanbuljs) that combines `jest` and `puppeteer` outputs, but certainly won't be wading into the stack any more than i have to on this PR.

File: packages/mermaid/test/clean.spec.ts
Comment (bollwyvl) [inline packages/mermaid/test/clean.spec.ts:25]: This would now error if the cleaned string is not valid XML... it's harder to verify that the browser would go into the "fallback" rendering mode (the most likely failure case).

Because of that, it's probably _not_ worth roundtripping inside the implementation, which is already pretty heavy.

File: galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png
Comment (krassowski) [reply galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png:1]: I pushed a fix for this using workaround from https://github.com/microsoft/playwright/issues/13486/#issuecomment-1112012053 (an upstream fix is apparently not planned).

File: galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts
Comment (bollwyvl) [inline galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts:33]: Perhaps this is something that should be hoisted up outside of a deeply-nested `utils`?

File: galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png
Comment (bollwyvl) [reply galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts-snapshots/mermaid-diagram-dark-09-requirement-jupyterlab-linux.png:1]: Thanks for the fix! Pretty weird its a non-priority upstream, as this hardly seems to that unique. As commented on the new file, perhaps this belongs someplace more prominent/easier to find?

**Comment by bollwyvl** [comment]
hooray all :green_circle:

File: galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts
Comment (krassowski) [reply galata/test/jupyterlab/notebook-mermaid-diagrams.test.ts:33]: Yes, we could add it as a galata helper in the future if it turns out useful elsewhere too. For now it's easier to do it here because this way we don't need to worry about it becoming a public API yet ;)

**Review by krassowski — APPROVED** [review]
Thank you @bollwyvl!

**Comment by bollwyvl** [comment]
Thanks. Alas, the wheels turn: since this was merged, https://github.com/mermaid-js/mermaid/releases/tag/v10.8.0 :crying_cat_face: 

I don't see any hard blockers/features that would _require_ going out, but will take a look, especially to kick the tires on the screenshot approach.

**Pull request merged**
